# Gotcha pipeline
library("Gotcha")

# --- Parameters --- #
input_dir = snakemake@params[["input_dir"]]
output_dir = snakemake@params[["output_dir"]]
singlecell = snakemake@params[["singlecell"]]
whitelist = snakemake@params[["whitelist"]]
primer = snakemake@params[["primer"]]
reverse_complement = snakemake@params['reverse_complement']
primed_max_mismatch = snakemake@params['primed_max_mismatch']
wt_sequence =  snakemake@params['wt_sequence']
mut_sequence = snakemake@params['mut_sequence']
mutation_start = snakemake@params['mut_start']
mutation_end = snakemake@params['mut_end']
read = snakemake@params['read']

# --- Analysis --- #
# 1. FastqSplit: if parallelization is available through a slurm cluster, this function allows to split the fastq files into chunks for parallel computing of by the BatchMutationCalling function.
print("Spliting fastq files...")
FastqSplit(path = input_dir, out = output_dir ,ncores = 12, reads = 500000) # reads = 1000,

# 2. Filter reads in the fastqs generated by gotcha, and filters them based on pre-determined base quality scores. Given a path to the folder containing the output of the FastqSplit function, it filters each fastq files for each chunk and stores them in a new folder named after the split suffix.
print("Filtering files fastq files...")
FastqFiltering(out = output_dir min.quality = 1, min.bases = 1, which.read = "R3", 
    read.region = c(mut_start:mut_end), ncores = 54)

# 3. Mut calling: Genotype each read and map it to the corresponding cell barcode
print("Applying mutation calling...")
BatchMutationCalling(out = output_dir,
                       whitelist.file.path = whitelist,
                       wt.max.mismatch = 0,
                       mut.max.mismatch = 0,
                       keep.raw.reads = F,
                       ncores = 10,
                       reverse.complement = reverse_complement,
                       testing = F,
                       which.read = read,
                       atac.barcodes = T,
                       primer.sequence = primer,
                       primed.max.mismatch = primed_max_mismatch,
                       atac.barcodes.file.path = singlecell,
                       wt.sequence =  wt_sequence,
                       mut.sequence = mut_sequence,
                       mutation.start = mutation_start,
                       mutation.end = mutation_end
                       )

# 4. Merge mut calls
#print("Merge mutation calls...")
#MergeMutationCalling(out = output_dir)
