<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2021-02-10" />

<title>Doublet Detection Vignette</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>





<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ATAC-DoubletDetector</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="ATAC-DoubletDetector.html">Doublet Detection Vignette</a>
    </li>
    <li>
      <a href="Doublet_Annotation_Pipeline.html">Doublet Annotation Vignette</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://www.ucarlab.com/">Ucar Lab</a>
</li>
<li>
  <a href="https://github.com/UcarLab/ATAC-DoubletDetector">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Doublet Detection Vignette</h1>
<h4 class="date">2021-02-10</h4>

</div>


<p><link rel="stylesheet" href="styles.css" type="text/css"> <link rel="stylesheet" href="academicicons/css/academicons.min.css"/></p>
<div id="user-guide" class="section level1">
<h1>User Guide</h1>
<div id="required-files" class="section level2">
<h2>Required Files</h2>
<p>The standard outputs from 10x’s <a href="https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/using/count"><code>cellranger-atac count</code></a> are used in this tutorial to detect the doublet cells. These are the position sorted BAM file and the CSV file containing the metadata for the single cell barcodes.</p>
<pre class="bash"><code>ls -1 example_data/pbmc1</code></pre>
<pre><code>#&gt; example_seurat_pbmc1.Rds
#&gt; possorted_bam.bam
#&gt; possorted_bam.bam.bai
#&gt; singlecell.csv</code></pre>
<p>The additional files rquired for the analysis are a TXT file with list of chromosomes to use for the analysis and a BED file with the repetitive regions for the chosen genome. The repetitive regions can be downloaded from <a href="https://github.com/UcarLab/ATAC-DoubletDetector/releases/download/v1.0-beta/RepeatFilterFiles.zip">here</a>.</p>
<pre class="bash"><code>head human_autosomes.txt</code></pre>
<pre><code>#&gt; chr1
#&gt; chr2
#&gt; chr3
#&gt; chr4
#&gt; chr5
#&gt; chr6
#&gt; chr7
#&gt; chr8
#&gt; chr9
#&gt; chr11</code></pre>
<pre class="bash"><code>head RepeatFilterFiles/blacklist_repeats_segdups_rmsk_hg38.bed</code></pre>
<pre><code>#&gt; chr1 10000   10468
#&gt; chr1 10627   10800
#&gt; chr1 10757   10997
#&gt; chr1 11225   11447
#&gt; chr1 11271   11448
#&gt; chr1 11283   11448
#&gt; chr1 19305   19443
#&gt; chr1 20828   20863
#&gt; chr1 30862   30959
#&gt; chr1 44835   44876</code></pre>
</div>
<div id="using-the-gui" class="section level2">
<h2>Using the GUI</h2>
<p>The initial page of the app is as follows.</p>
<p><img src="screenshots/Step1.png" width="603" /></p>
<p>The required files can be found from the browser and chosen.</p>
<p><img src="screenshots/Step2_Data_Fill.png" width="645" /></p>
<p>The shell command is formed with the appopriate locations and can be used by paste to a shell terminal.</p>
<pre class="sh"><code>&quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/ATACDoubletDetector.sh&quot; --bambc CB --bcidx 0 --cellidx 8 --iscellidx 9 &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/example_data/pbmc1/possorted_bam.bam&quot; &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/example_data/pbmc1/singlecell.csv&quot; &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/human_autosomes.txt&quot; &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/RepeatFilterFiles/blacklist_repeats_segdups_rmsk_hg38.bed&quot; &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/example_output/pbmc1/&quot; &quot;/Users/erogla/Desktop/ATAC-DoubletDetector_Release_12-4-2020/&quot;
</code></pre>
<p>ATAC-DoubletDetector can be run from the app by clicking <code>Run Shell Script</code>. The initial part of the program will start working by looking at the regions that are repeated more than twice for all of the cells.</p>
<p><img src="screenshots/Step3_Start.png" width="401" /></p>
<p><img src="screenshots/Step4_Read_Count_Finish.png" width="401" /></p>
<p>The next step is running the python code that detects the cells that are doublets based on the number of regions that are repeated more than twice. When finished, the program can be closed by clicking <code>Finish</code>.</p>
<p><img src="screenshots/Step5_DoubletChosen.png" width="401" /></p>
<p>The output files will be in the chosen output directory and look as follows.</p>
<pre class="bash"><code>ls example_output/pbmc1</code></pre>
<pre><code>#&gt; DoubletBarcodes_01.txt
#&gt; DoubletCellIds_01.txt
#&gt; DoubletProbabilities.txt
#&gt; DoubletSummary.txt
#&gt; OverlapSummary.txt
#&gt; Overlaps.txt
#&gt; RunTime.txt
#&gt; StatSummary.txt</code></pre>
<p>The summary of the program will look like the following:</p>
<pre class="bash"><code>cat example_output/pbmc1/DoubletSummary.txt</code></pre>
<pre><code>#&gt; Number of Cells  6143
#&gt; Number of Merged Regions 31375
#&gt; Number of Doublets   431
#&gt; Doublet Percent  7.016115904281296</code></pre>
<p>The ids of the cells that are identified as doublets are the following. These can be removed from a Seurat object in the downstream analyses after following the <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">Signac Pipeline</a>.</p>
<pre class="bash"><code>head example_output/pbmc1/DoubletCellIds_01.txt</code></pre>
<pre><code>#&gt; _cell_1000
#&gt; _cell_1003
#&gt; _cell_1013
#&gt; _cell_1036
#&gt; _cell_1038
#&gt; _cell_1044
#&gt; _cell_1063
#&gt; _cell_1064
#&gt; _cell_1070
#&gt; _cell_1083</code></pre>
<p>First required packages for the pipeline are loaded in.</p>
<pre class="r"><code>library(Seurat)
library(Signac)
library(dplyr)
library(tidyr)
library(textclean)
library(ggplot2)
library(stringr)</code></pre>
<p>Ids of the doublet cells are read in and have the following format</p>
<pre class="r"><code>doublets &lt;- read.table(&quot;example_output/pbmc1/DoubletCellIds_01.txt&quot;) %&gt;% t() %&gt;% as.vector()

head(doublets)</code></pre>
<pre><code>#&gt; [1] &quot;_cell_1000&quot; &quot;_cell_1003&quot; &quot;_cell_1013&quot; &quot;_cell_1036&quot; &quot;_cell_1038&quot;
#&gt; [6] &quot;_cell_1044&quot;</code></pre>
<p>The Seurat object processed by the <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">Signac Pipeline</a> is loaded.</p>
<pre class="r"><code>pbmc.atac &lt;- readRDS(file = &quot;example_data/pbmc1/example_seurat_pbmc1.Rds&quot;)

pbmc.atac</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 174800 features across 6143 samples within 1 assay 
#&gt; Active assay: ATAC (174800 features, 174800 variable features)
#&gt;  2 dimensional reductions calculated: svd, umap</code></pre>
<p>Cells in the Seurat object must have the same format as the doublets ids.</p>
<p>Get all of the cells in the seurat object.</p>
<pre class="r"><code>cells &lt;- Cells(pbmc.atac)

head(cells)</code></pre>
<pre><code>#&gt; [1] &quot;_cell_0&quot;    &quot;_cell_1&quot;    &quot;_cell_10&quot;   &quot;_cell_100&quot;  &quot;_cell_1000&quot;
#&gt; [6] &quot;_cell_1001&quot;</code></pre>
<p>Add the doublet information for the cells in the Seurat object.</p>
<pre class="r"><code>doublet.metadata &lt;- rep(&quot;singlet&quot;, length(cells))
names(doublet.metadata) &lt;- cells
doublet.metadata[doublets] &lt;- &quot;doublet&quot;

pbmc.atac &lt;- AddMetaData(pbmc.atac, doublet.metadata, col.name = &quot;doublet&quot;)

DimPlot(pbmc.atac, group.by = &quot;doublet&quot;, pt.size = 1, 
        cols = c(&quot;singlet&quot; = &quot;grey&quot;, &quot;doublet&quot; = &quot;purple&quot;), 
        order = c(&quot;doublet&quot;, &quot;singlet&quot;))</code></pre>
<p><img src="ATAC-DoubletDetector_files/figure-html/unnamed-chunk-17-1.png" width="480" /></p>
<p>Remove the doublet cells from the Seurat object and perform the downstream analyses.</p>
<pre class="r"><code>pbmc.atac &lt;- subset(pbmc.atac, doublet == &quot;singlet&quot;)

pbmc.atac</code></pre>
<pre><code>#&gt; An object of class Seurat 
#&gt; 174800 features across 5712 samples within 1 assay 
#&gt; Active assay: ATAC (174800 features, 174800 variable features)
#&gt;  2 dimensional reductions calculated: svd, umap</code></pre>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
